feat: Allow users to explicitly configure universe domain.

ohmayr opened the pull request:
Fixes #1900 🦕
Fixes #1866

ohmayr committed:
add support for universe domain

ohmayr committed:
add test cases for universe domain

ohmayr committed:
update tests for read environment variables

ohmayr committed:
update tests for api endpoint

ohmayr committed:
fix variable name

ohmayr committed:
move check for empty universe and update tests

ohmayr committed:
update universe property for async

ohmayr committed:
clean up tests

ohmayr committed:
use the correct endpoint in tests

ohmayr committed:
fix template name in tests

ohmayr committed:
update tests to use DEFAULT_ENDPOINT_TEMPLATE

ohmayr committed:
missing format for string

ohmayr committed:
revert mtls_endpoint function testcases

ohmayr committed:
fix get_mtls_endpoint_and_cert_source tests

ohmayr committed:
ctestuniverse domain property

ohmayr committed:
add validation step before sending the request

ohmayr committed:
change _transport to transport

ohmayr committed:
change _transport to transport for async

ohmayr committed:
change credentials to _credentials

ohmayr committed:
Implement AnonymousCredentialsWithUniverseDomain for testing

ohmayr committed:
fix import error for google auth creds

ohmayr committed:
set default value for universe in creds

ohmayr committed:
change universe domain property of credentials

ohmayr committed:
add property of universe domain to AnonymousCredentialsWithUniverseDomain

ohmayr committed:
fix comma, update client.py

ohmayr committed:
update credentials in conftest.py

ohmayr committed:
move AnonymousCredentials to conftest and templates

ohmayr committed:
keep credentials for grpc if channel is passed

ohmayr committed:
only ignore credentials if credentials are not passed with channel

ohmayr committed:
wrap universe validation into a function

ohmayr committed:
add universe validation method to async

ohmayr committed:
update contract for validation method

ohmayr committed:
refactor code

ohmayr committed:
fix whitespace

ohmayr committed:
add test case for validate universe and change method to static

ohmayr committed:
fix syntax of test case string

ohmayr committed:
remove self from method call

ohmayr committed:
remove self from method call in async mixins

ohmayr committed:
fix mixins validation

ohmayr committed:
fix value error string for validation exception

ohmayr committed:
fix universe property in async mixins

ohmayr committed:
remove universe validation method from async

ohmayr committed:
avoid always ignoring credentials for async client with transport

ohmayr committed:
pass credentials to intercepted echo

ohmayr committed:
cleanup whitespace and revert unnecessary changes

ohmayr committed:
add integration tests for universe domain

ohmayr committed:
fix style check for tests

ohmayr committed:
remove client options from echo localhost

ohmayr committed:
comment out integration tests for universe

ohmayr committed:
fix whitespace

ohmayr committed:
validate universe if not already validated

ohmayr committed:
refactor universe validation step

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -131,12 +131,17 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
 
         return api_endpoint.replace(".googleapis.com", ".mtls.googleapis.com")
 
+    # Note: DEFAULT_ENDPOINT is deprecated. Use DEFAULT_ENDPOINT_TEMPLATE instead.
```
Deprecation note for `DEFAULT_ENDPOINT`. This should no longer be used anywhere in the code except for `_get_mtls_endpoint_and_cert_source` tests which will eventually be removed once we clean up the code.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:
```diff
@@ -133,8 +133,8 @@ class {{ service.name }}GrpcTransport({{ service.name }}Transport):
-            # Ignore credentials if a channel was passed.
-            credentials = False
+            # Ignore credentials if a channel was passed without credentials.
+            credentials = False if not credentials else credentials
```
This wasn't a part of the design so adding some clarification. Validation step (for universe domain) requires us to access the universe domain from the credentials, hence we can no longer ignore credentials if a channel is explicitly passed in (mainly for testing purposes).
i.e. the condition:
```python3
if self.universe_domain != self.transport._credentials.universe_domain
```
fails without this change.

vchudnov-g replied:
Would this be a breaking change? Before, if the user was passing a channel with credentials, we were clearing the credentials. Now we're not. So that will affect the behavior of this constructor, it would seem.

One idea for not changing the code is that when we try to access the universe domain from the credentials, we first check to see whether the credentials are set, and if they're not, consider the universe domain to be the empty string or `None`. Would this help the testing scenario?

ohmayr replied:
Addressed!

ohmayr commented on gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2:
```diff
@@ -6,7 +6,7 @@
 def test_{{ method_name }}(request_type, transport: str = 'grpc'):
     client = {{ service.client_name }}(
-        credentials=ga_credentials.AnonymousCredentials(),
+        credentials=AnonymousCredentialsWithUniverseDomain(),
```
Since validation step is a part of each RPC call, we need to pass in credentials with universe domain property to each test. Otherwise, they will complain about missing `universe_domain` property in the validation step.

ohmayr committed:
add cleanup comment and fix whitespace

parthea commented on tests/system/test_universe_domain.py:
```diff
@@ -0,0 +1,20 @@
+import pytest
+
+# Note(TODO): Uncomment these test cases once showcase library is updated
```
Please fix. We can update the minimum version of `google-api-core` if necessary.

ohmayr replied:
addressed!

parthea reviewed (commented):
LGTM. Please update the version of `google-api-core` if required. Note, we still support `google-api-core` 1.x so we will need to cherry-pick the necessary change in the `google-api-core` 1.x branch.

ohmayr committed:
uncomment universe domain tests

ohmayr committed:
raise a warning if universe_domain does not exist

ohmayr committed:
use fstring and update tests to check property exists

ohmayr committed:
check for attribute in object instance

ohmayr committed:
raise exception in else clause

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:
```diff
@@ -133,8 +133,8 @@ class {{ service.name }}GrpcTransport({{ service.name }}Transport):
-            # Ignore credentials if a channel was passed.
-            credentials = False
+            # Ignore credentials if a channel was passed without credentials.
+            credentials = False if not credentials else credentials
```
```suggestion
            credentials = credentials or False
```
if you really need a Boolean value, or even better, just delete the line entirely: if `credentials` is not passed in, it is `None` so it evaluates as `False`. In other words, if you don't need a strict Boolean type, this line is a no-op and can be removed. But again, it changes the behavior of this method.

ohmayr replied:
addressed!

vchudnov-g reviewed (commented):
Mostly minor comments, but two non-trivial ones
- we should determine whether the change to the channels+credentials check is breaking
- suggestions for consolidating the check for valid universe domain

ohmayr committed:
set correct value for client in else case

vchudnov-g commented:
@ohmayr Please re-assign back to me once it's ready for the next round of review.

ohmayr committed:
add test coverage for universe domain attribute not existing

ohmayr committed:
use client options object in instead of mock in test

ohmayr committed:
change Client_Options to ClientOptions

ohmayr committed:
warning raised in and not error in test case

ohmayr committed:
remove trailing comma

ohmayr committed:
only delete attribute if it exists

ohmayr committed:
update nox sessions to avoid testing universe for ads

ohmayr committed:
use environment variable to ignore file

ohmayr committed:
ignore test universe as an argument to pytest command

ohmayr committed:
update fstring

ohmayr committed:
use path instead of f string

ohmayr committed:
convert path to string

ohmayr committed:
set default value of env to empty dict

ohmayr committed:
fix directory path

ohmayr committed:
update path

ohmayr committed:
update tests for universe domain

ohmayr committed:
update the error message to match regex pattern

ohmayr committed:
update err.match to string comparison

ohmayr committed:
compare only the value

ohmayr committed:
minor fixes

ohmayr committed:
make validation variable private

ohmayr committed:
update test case to use object to support len operator

ohmayr committed:
update _get_universe_domain function

ohmayr committed:
fix parenthesis

ohmayr committed:
revert _get_universe_domain function

ohmayr committed:
only validate universe if credentials exist

ohmayr committed:
remove staticmethod decorator

ohmayr commented on tests/integration/goldens/logging/google/cloud/logging_v2/services/logging_service_v2/client.py:
```diff
@@ -308,11 +312,12 @@ def _read_environment_variables():
         if use_mtls_endpoint not in ("auto", "never", "always"):
             raise MutualTLSChannelError("Environment variable `GOOGLE_API_USE_MTLS_ENDPOINT` must be `never`, `auto` or `always`")
-        return use_client_cert == "true", use_mtls_endpoint
+        return use_client_cert == "true", use_mtls_endpoint, universe_domain_env
```
shouldn't we keep the logic for figuring out the `mtls_endpoint` in the `get_api_endpoint function`? The purpose of this function is just to read variables as the name suggests i.e. `read_environment_variables`.

vchudnov-g replied:
Sorry, I forgot to delete this comment before sending out the review. Ignore.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        Returns:
+            str: The universe domain to be used by the client.
+        """
+        universe_domain = {{ service.client_name }}.GOOGLE_DEFAULT_UNIVERSE
+        if client_universe_domain is not None:
+            universe_domain = client_universe_domain
+        elif universe_domain_env is not None:
+            universe_domain = universe_domain_env
```
This is a much concise alternative but I run into some issues when I make this change:
- `""` is treated as None in the `or` statement and the statement will end up evaluating to the default GDU value. As a result, the below exception for checking the empty string will never be raised (the exception is part of the design.)
- If I were to split the `or` statement to two, the `len` check on the None type would lead to an error.
- If I do the error checking before assigning the `universe_domain` value for each individual parameter, then I end up raising an exception for the following scenario i.e. `client_universe_domain="foo.com"` and `universe_domain_env=""`. In this case, `client_universe_domain` is a valid universe and we shouldn't bother if a universe is configured (even incorrectly) via the environment variable.

There are obviously workarounds to make this work but that seem to defeat the purpose of what's proposed here. 

Feel free to suggest more on this!

vchudnov-g replied:
Oh, good point.

I think this should work, though it's less clear than the non-working version

```suggestion
EMPTY_TOKEN=-9999 # since empty strings are falsy, but we want to distinguish them from None
universe_domain = (((client_universe_domain is not None) and (client_universe_domain or EMPTY_TOKEN)) or
                   ((universe_domain_env is not None) and (universe_domain_env or EMPTY_TOKEN)) or
                   {{ service.client_name }}.GOOGLE_DEFAULT_UNIVERSE)
 if universe_domain == EMPTY_TOKEN or len(universe_domain.strip()) == 0:
     raise ValueError(...)
```

vchudnov-g replied:
Your call on which of the two alternatives is easier to follow. I slightly prefer my version, but there's a tradeoff between conciseness and thinking through the code---not a big burden either way in this case.

ohmayr replied:
I think let's keep it as is since it looks easier to follow through for someone who might not be familiar with what's happening here.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -131,12 +131,17 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
 
+    DEFAULT_ENDPOINT_TEMPLATE = {% if service.host %}"{{ service.host.replace("googleapis.com", "{UNIVERSE_DOMAIN}") }}"{% else %}None{% endif %}
+    
+    GOOGLE_DEFAULT_UNIVERSE = "googleapis.com"
```
`GOOGLE_DEFAULT_UNIVERSE` can definitely be moved to `api_core` but not `DEFAULT_ENDOINT_TEMPLATE` since the value for it is dynamically resolved.

Note that we're using an older version of `api_core` for `python 3.7`. As a result we'll have to have the check `hasattr(Client_options, "universe_domain")` every time we use code from `api_core` in gapic that we recently add.

vchudnov-g replied:
So then we would not have custom universes for Python 3.7? Does that go against the requirements of this project, or is that acceptable?

If that would be a problem, you can keep the code as is but add a TODO outlining two options I see (do you see others?):
*  when we stop supporting Python 3.7, we can move this symbol to api_core; or
* we could add the symbol to the latest version of `api_core` and at the same time:
  *  take the version of `api_core` we support for 3.7, make a branch for that, and there only add what we need for this feature. But we'd have to be clear on how to release it and on making sure we pin the dependency correctly. 


(Probably not worth investing the time doing all that coding work at the moment, which is why I'm suggesting a TODO if we need to support the feature in 3.7)

parthea replied:
> So then we would not have custom universes for Python 3.7? Does that go against the requirements of this project, or is that acceptable?

The issue is not specific to python 3.7 but rather whether the user has an older version of `google-api-core` installed and wants to use the new feature which requires a newer version of `google-api-core`

Possible solution : If the user sets `universe_domain` and has an older version of `google-api-core` installed, we can raise an error that a newer version of `google-api-core`  is required (we should provide the users with specific minimum version).

vchudnov-g replied:
As per our off-line discussion, we'll defer refactoring some functionality into `google-api-core` until after this PR is in.

vchudnov-g replied:
And I withdraw my original suggestion here. The string substitution is happening at the Jinjha template level, so what I suggested does not make sense.

That said, I would strongly encourage taking this business logic out of the templates and into the generator, but that should probably be part of a larger refactoring.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            raise ValueError("Universe Domain cannot be an empty string.")
+        return universe_domain
+    
+    @staticmethod
```
1. and 2. can definitely be done (having workarounds in place so we don't break code if an older version of `api_core` is being used). [I'm working on this.]

3. This is decided in the design for TPC and needs to be consistent across platforms i.e.
- We should wait for validating the universe domain until the first call to an API is made.

Also, even if a user provides an incorrect universe domain value that does not match the one in credentials, they should be able to initialize a client and be able to read the value for universe domain in it. The following should work:

```python3

client = EchoClient(ClientOptions{"universe_domain":"foo.com"})
client.universe_domain

vchudnov-g replied:
Hmmm, OK. I'll have to refresh myself on the motivation for that design decision. 

But OK, if that's the feature requirement, we can't do 3. (And I'm not seeing any other similar lazy initialization we can lump together with this one.)

vchudnov-g replied:
Also:

> (having workarounds in place so we don't break code if an older version of api_core is being used). 

If the workarounds make the generated code less clear or less readable, we may just have to make do with a TODO for now saying when we can do what simplifications.

ohmayr replied:
I've addressed this by implementing [1.] and [2.] in the following fashion:
- I have exactly what you've proposed except I've moved the logic for raising the exception to a new private function `compare_universes` which is a place holder for `api_core.validate_universe_domain`. This way, when we eventually plan to move code to `api_core`, we can just move this function directly without requiring much refactor.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -345,7 +381,17 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            str: The universe domain used
+                by the client instance.
+        """
+        return self._universe_domain
```
Users should be able to access the value that they've set even if it's incorrect? The validation will be delayed until an API call is made (as part of the design).

vchudnov-g replied:
ACK.

ohmayr committed:
pass in credentials to client for validation unit tests

ohmayr committed:
ignore failing validation test for grpc

ohmayr committed:
add link to code cleanup issue in test template

ohmayr committed:
add comment for copying defaults from sync

ohmayr commented on gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:
```diff
@@ -78,13 +78,29 @@ from google.iam.v1 import policy_pb2  # type: ignore
     return b"cert bytes", b"key bytes"
 
-
+{#TODO: Remove this function as part of cleanup when DEFAULT_ENDPOINT is no longer used.#}
```
addressed!

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:
```diff
@@ -54,8 +54,11 @@ class {{ service.async_client_name }}:
 
     _client: {{ service.client_name }}
 
+    # Note: DEFAULT_ENDPOINT is deprecated. Use DEFAULT_ENDPOINT_TEMPLATE instead.
```
addressed!

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         Args:
             api_override (str): The API endpoint override. If specified, this is always the return value of this function.
             client_cert_source (bytes): The client certificate source used by the client.
+            universe_domain (str): The universe domain used by the client.
```
addressed!

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
             str: The API endpoint to be used by the client.
         """
-
+        use_mtls_endpoint = use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source)
```
addressed!

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            the universe domain in the credentials.
+        """
+        if client_universe != credentials_universe:
+            raise ValueError("The configured universe domain (" + client_universe + ") does not match the universe domain found in the credentials (" + credentials_universe + "). If you haven't configured the universe domain explicitly, `googleapis.com` is the default.")
```
addressed!

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -581,6 +631,10 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         )
 
+        # Validate the universe domain.
+        self._validate_universe_domain(self.universe_domain, self.transport._credentials.universe_domain) if not self.is_universe_domain_valid else None
```
addressed!

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        return universe_domain
+    
+    @staticmethod
+    def _validate_universe_domain(client_universe, credentials_universe):
```
addressed.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -345,7 +381,17 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 by the client instance.
         """
-        return self._api_endpoint    
+        return self._api_endpoint
```
addressed!

ohmayr committed:
refactor _validate_universe_domain

ohmayr committed:
make compare_universes a static method

ohmayr committed:
fix type in test case

ohmayr committed:
swap universe domain in test case

ohmayr committed:
make variables private

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            the universe domain in the credentials.
+        """
+        if client_universe != credentials_universe:
+            raise ValueError("The configured universe domain (" + client_universe + ") does not match the universe domain found in the credentials (" + credentials_universe + "). If you haven't configured the universe domain explicitly, `googleapis.com` is the default.")
```
addressed.

ohmayr committed:
missed _ in AnonymousCredentialsWIthUniverseDomain

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,94 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            client_universe (str): The universe domain configured via the client options.
+            credentials_universe (str): The universe domain in the credentials.
+
+        Returns:
+            bool: True if valid universe, else raise a "ValueError" exception.
```
```suggestion
        Returns:
            bool: Returns True if valid universe domain is valid, otherwise False.
        Raises:
            ValueError: If universe domain is an empty string.
```

ohmayr replied:
addressed.

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,94 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
 
         Args:
-            api_override (str): The API endpoint override. If specified, this is always the return value of this function.
+            api_override (str): The API endpoint override. If specified, this is always the return value of this function
```
nit: line wrap (cf https://google.github.io/styleguide/pyguide.html#32-line-length)

ohmayr replied:
addressed.

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,94 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
             api_endpoint = api_override
         elif use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source):
+            if universe_domain != {{ service.client_name }}._DEFAULT_UNIVERSE:
+                raise MutualTLSChannelError("MTLS is not supported in any universe other than googleapis.com.")
```
```suggestion
                raise MutualTLSChannelError(f"MTLS is not supported in any universe other than { {{ service.client_name }}._DEFAULT_UNIVERSE} ")
```

ohmayr replied:
addressed.

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,94 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            credentials_universe (str): The universe domain in the credentials.
+
+        Returns:
+            bool: True if valid universe, else raise a "ValueError" exception.
```
```suggestion
            bool: True if the universes match
            
          Raises: 
            ValueError: when the universes do not match.
```

ohmayr replied:
addressed.

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,94 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        Returns:
+            bool: True if valid universe, else raise a "ValueError" exception.
+        """
+        self._is_universe_domain_valid = {{ service.client_name }}._compare_universes(self.universe_domain, self.transport._credentials.universe_domain)
```
```suggestion
        self._is_universe_domain_valid = (self._is_universe_domain_valid or
                                         {{ service.client_name }}._compare_universes(self.universe_domain, self.transport._credentials.universe_domain))
```

ohmayr replied:
addressed.

ohmayr committed:
move credentials check inside function and update contracts

ohmayr committed:
update tests

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -402,9 +466,19 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        try:
+            universe_domain_opt = getattr(self._client_options, 'universe_domain')
+        except AttributeError:
+            warnings.warn("Attribute universe_domain does not exist in self._client_options. Setting universe_domain_opt to None.")
```
nit: I wonder whether we need this warning. If the user is passing in a dict that doesn't have this key, do we want to warn? Seems a bit annoying....and this behavior of "if it's not present, behave as though it were set to None" seems like what people would expect.

parthea replied:
Perhaps we could just let the error bubble up in the case that the user passes a dictionary with unsupported attributes

vchudnov-g replied:
This is passing a dictionary with an attribute missing, and it seems to me this attribute should be semantically optional (if not present, we default to it being set to `None` and use the GDU. So catching the exception seems fine; it's just the warning that I would remove.

vchudnov-g replied:
Going with silently defaulting to `None`. Please speak up if there are objections.

parthea replied:
You're right @vchudnov-g . No warning is needed here. I was thinking of a different code path https://github.com/googleapis/gapic-generator-python/blob/414e3cc5e4246b7bf86df6500f68bab443e50c32/gapic/templates/%25namespace/%25name_%25version/%25sub/services/%25service/client.py.j2#L400

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -402,9 +466,19 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
 
-        self._use_client_cert, self._use_mtls_endpoint = {{ service.client_name }}._read_environment_variables()
+        try:
+            universe_domain_opt = getattr(self._client_options, 'universe_domain')
```
In the `__init__` docstring above, the section for `client_options` says " It won't take effect if a ``transport`` instance is provided." That should be updated, and the description should have a third clause explaining the new ``universe_domain`` property.

ohmayr replied:
addressed.

vchudnov-g replied:
Replicated docstring for async client (and updated other parts of the docstring that were out of date).

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -581,6 +655,9 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         )
 
+        # Validate the universe domain.
+        self.transport._credentials and self._validate_universe_domain()
```
Let's move the `self.transport._credentials` check into `_validate_universe_domain`, with something in that method like

```
self._is_universe_domain_valid = ( self.transport._credentials is None or
                                   ConfigServiceV2Client._compare_universes(self.universe_domain, self.transport._credentials.universe_domain)
```


and then this line here becomes just the function call

ohmayr replied:
addressed.

vchudnov-g reviewed (commented):
Replying with some comments after looking at the generated `client` files. I'll keep looking at the other files, but feel free to start acting on these comments, (ping me again when the updated code is stable for review)

ohmayr committed:
add test coverage

ohmayr committed:
fix client name in test case

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/_async_mixins.py.j2:
```diff
@@ -47,6 +47,9 @@
         )
 
+        # Validate the universe domain.
+        self.transport._credentials and self._client._validate_universe_domain()
```
addressed.

ohmayr commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -402,9 +466,19 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        try:
+            universe_domain_opt = getattr(self._client_options, 'universe_domain')
+        except AttributeError:
+            warnings.warn("Attribute universe_domain does not exist in self._client_options. Setting universe_domain_opt to None.")
```
addressed.

ohmayr commented on tests/integration/goldens/redis/google/cloud/redis_v1/services/cloud_redis/client.py:
```diff
@@ -359,37 +364,95 @@ def _get_client_cert_source(provided_cert_source, use_cert_flag):
+
+        Returns:
+            str: The universe domain used
+                by the client instance.
```
addressed.

vchudnov-g commented:
BTW, for future reference: we decided in a meeting earlier today that we would not yet port parts of this implementation to `api_core`, as I had suggested. We will do that as a non-breaking follow-up set of changes, since other issues are prevent quick changes to api_core at the moment and we want to release this functionality soon.

ohmayr committed:
fix transport name in tests

ohmayr committed:
minor updates to the universe feature

ohmayr committed:
add missing period

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -131,12 +131,17 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         DEFAULT_ENDPOINT
     )
 
+    DEFAULT_ENDPOINT_TEMPLATE = {% if service.host %}"{{ service.host.replace("googleapis.com", "{UNIVERSE_DOMAIN}") }}"{% else %}None{% endif %}
+    
+    _DEFAULT_UNIVERSE = "googleapis.com"
```
Suggest not hard-coding the string replacement:

```suggestion
    _DEFAULT_UNIVERSE = "googleapis.com"
    DEFAULT_ENDPOINT_TEMPLATE = {% if service.host %}"{{ service.host.replace(_DEFAULT_UNIVERSE, "{UNIVERSE_DOMAIN}") }}"{% else %}None{% endif %}
    

```

vchudnov-g replied:
Done

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,103 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        """
+        default_universe = {{ service.client_name }}._DEFAULT_UNIVERSE
+        if client_universe != credentials_universe:
+            raise ValueError(f"The configured universe domain ({client_universe}) does not match the universe domain found in the credentials ({credentials_universe}). If you haven't configured the universe domain explicitly, `{default_universe}` is the default.")
```
Consistent quoting of universes:
```suggestion
            raise ValueError(f"The configured universe domain `({client_universe})` does not match the universe domain found in the credentials `({credentials_universe})`. If you haven't configured the universe domain explicitly, `{default_universe}` is the default.")
```

vchudnov-g replied:
NM. They're in parentheses, so that's fine.

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,103 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+
+        Returns:
+            bool: Returns True if universes match.
+        Raises:
```
```suggestion

        Raises:
```

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,103 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        
+        Returns:
+            bool: Returns True if universe domain is valid, otherwise False.
+        Raises:
```
```suggestion

        Raises:
```

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,103 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            credentials_universe (str): The universe domain in the credentials.
+
+        Returns:
+            bool: Returns True if universes match.
```
```suggestion
            bool:  True if universes match.
```

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,103 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+            the universe domain in the credentials.
+        
+        Returns:
+            bool: Returns True if universe domain is valid, otherwise False.
```
```suggestion
            bool: True if universe domain is valid, otherwise False.
```

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:
```diff
@@ -54,8 +54,12 @@ class {{ service.async_client_name }}:
+    # Note: DEFAULT_ENDPOINT is deprecated. Use DEFAULT_ENDPOINT_TEMPLATE instead.
     DEFAULT_ENDPOINT = {{ service.client_name }}.DEFAULT_ENDPOINT
     DEFAULT_MTLS_ENDPOINT = {{ service.client_name }}.DEFAULT_MTLS_ENDPOINT
+    DEFAULT_ENDPOINT_TEMPLATE = {{ service.client_name }}.DEFAULT_ENDPOINT_TEMPLATE
```
Does this symbol really need to be public? If not, let's underscore it  (here and in the sync_client)

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:
```diff
@@ -133,7 +133,7 @@ class {{ service.name }}GrpcTransport({{ service.name }}Transport):
 
         if channel:
-            # Ignore credentials if a channel was passed.
+            # Ignore credentials if a channel was passed
```
```suggestion
            # Ignore credentials if a channel was passed.
```

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2:
```diff
@@ -178,7 +178,7 @@ class {{ service.grpc_asyncio_transport_name }}({{ service.name }}Transport):
 
         if channel:
-            # Ignore credentials if a channel was passed.
+            # Ignore credentials if a channel was passed
```
```suggestion
            # Ignore credentials if a channel was passed.
```

vchudnov-g reviewed (commented):
Some more incremental comments.

vchudnov-g commented on gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:
```diff
@@ -556,58 +614,63 @@ def test_{{ service.client_name|snake_case }}_get_mtls_endpoint_and_cert_source(
-        options = client_options.ClientOptions(client_cert_source=mock_client_cert_source, api_endpoint=mock_api_endpoint)
-        client = client_class(client_options=options, credentials=ga_credentials.AnonymousCredentials())
-        assert client.api_endpoint == mock_api_endpoint
+        options = client_options.ClientOptions(client_cert_source=mock_client_cert_source, api_endpoint=api_override)
```
It seems like we're testing combinations of 3 paramters, `GOOGLE_API_USE_CLIENT_CERTIFICATE` `(true` or `false`), `GOOGLE_API_USE_MTLS_ENDPOINT` (`never`, `always`, `auto`) and `ClientOptions.api_endpoint` ( set or unset) for 2×3×2=12 combinations. We should list and test them systematically. For those combinations that have the same behavior, we could just choose one representative test case and add a comment about the other cases expected to behave the same way

vchudnov-g commented on gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:
```diff
@@ -556,58 +614,63 @@ def test_{{ service.client_name|snake_case }}_get_mtls_endpoint_and_cert_source(
+    assert client.api_endpoint == (mock_endpoint if universe_exists else default_endpoint)
+    assert client.universe_domain == (mock_universe if universe_exists else default_universe)
+
+    # If ClientOptions does not have a universe domain attribute and GOOGLE_API_USE_MTLS_ENDPOINT="never",
```
We could split this into a separate test function, and here too it seems like we have several options to test. Maybe the new test function should be like the current one, except that it explicitly deletes `"universe_domain"` in all test cases. Or we choose one or a few representative test cases? WDYT? I don't want to go overboard, so check me on this.

parthea replied:
Please add a todo comment for the verbose test case.

vchudnov-g reviewed (commented):
Some more comments. Looking good!

ohmayr committed:
some minor updates

ohmayr committed:
revert fixture params

vchudnov-g committed:
fix: minor tweaks suggested udring code review

This does not include the testing suggestions, which we'll implement
later, possibly in a follow-up PR.

vchudnov-g committed:
fix: fix universe template creation

vchudnov-g committed:
chore: update test files for foregoing changes

vchudnov-g committed:
chore: update test files and golden files

vchudnov-g committed:
fix: decorate with @staticmethod in order to pass mypy checks

vchudnov-g committed:
test: Fix Showcase tests and corrected logic.

More changes need to occur for the system test.

vchudnov-g commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:
```diff
@@ -133,8 +134,6 @@ class {{ service.name }}GrpcTransport({{ service.name }}Transport):
             warnings.warn("client_cert_source is deprecated", DeprecationWarning)
 
         if channel:
-            # Ignore credentials if a channel was passed.
```
@ohmayr I wound up removing this again, as you had originally. I convinced myself that this will not break the API surface, and we do need it for testing unless we make major changes to the architecture.

vchudnov-g replied:
Update: now I'm putting it back, because it looks like removing it screwed u pother tests. I think we could make a case for correct behavior either way, but keeping it in is fine. And I think I was able to finesse the changes in the new test from leaving it in.

vchudnov-g committed:
chore: update goldens

vchudnov-g committed:
fix: initialize endpoint on client

vchudnov-g committed:
fix: minor simplification, update goldens

vchudnov-g committed:
fix: more test anhancements

vchudnov-g committed:
fix: cosmetic+goldens

vchudnov-g committed:
fix: remove extraneous print statements used in dev

vchudnov-g committed:
nit: fix comment cruft

vchudnov-g committed:
fix: lint source files

vchudnov-g committed:
fix: update signature for compare universes

forgot to include this when I was separatng local debug changes from
actual code changes

vchudnov-g committed:
fix mypy failures

vchudnov-g committed:
core: update goldens

vchudnov-g committed:
fix: restore seting credentials to False when a channel is provided

We must clarify that we don't use the credentials at all in that case.

vchudnov-g committed:
fix: fix ADC presubmit error: pass intended transport in test

vchudnov-g committed:
disable test case depending on api-core version unsupported in Py3.7

vchudnov-g committed:
fix: clarify test-skipping comment

vchudnov-g committed:
fix: fix api-core condition gating test case

vchudnov-g commented:
Now that you're back, @ohmayr , LMK what you think of my recent commits. IMO, we're ready to merge this, but I certainly want an extra mind verifying my work.

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:
```diff
@@ -172,22 +184,42 @@ class {{ service.async_client_name }}:
-                environment variable can also be used to override the endpoint:
+                {% if 'rest' in opts.transport and not opts.rest_numeric_enums %}
+                NOTE: "rest" transport functionality is currently in a
+                beta state (preview). We welcome your feedback via an
```
TODO: file a bug to remove the beta (preview) comment

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,110 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         return api_endpoint
 
+    @staticmethod
+    def _get_universe_domain(client_universe_domain, universe_domain_env):
```
Add type hints for `client_universe_domain` and `client_universe_domain` as well as type hints for the return value.

parthea commented on tests/system/test_universe_domain.py:
```diff
@@ -0,0 +1,94 @@
+    assert response.content == "Universe validation succeeded!"
+
+
+# What we'd like to do for the following test is:
+#    @pytest.mark.parametrize("channel_creator", [grpc.insecure_channel])
+#    @pytest.mark.parametrize("transport_name",["grpc","rest"])
+#    @pytest.mark.parametrize("transport_endpoint", ["showcase.googleapis.com", "localhost:7469" ])
+#    @pytest.mark.parametrize("credential_universe",["showcase.googleapis.com", "localhost:7469"])
```
Remove commented code. Add a comment that reads like a comment instead of code. Applies throughout.

vchudnov-g replied:
Drive-through comment: the comment would be about not getting this code to work, so including the code seems the most concise way. Maybe we can expand the preamble, saying the this attempted parametrization stacking was not working.

parthea replied:
@ohmayr , Please could you take a look to see if we can achieve the same result using `@pytest.mark.parametrize` similar to how we do it here? Let me know if you'd like to pair program on it

https://github.com/googleapis/gapic-generator-python/blob/2e86a654896054f1ac4b6833127bb095d2bf5b90/tests/integration/goldens/asset/tests/unit/gapic/asset_v1/test_asset_service.py#L223-L227

parthea commented on tests/system/conftest.py:
```diff
@@ -107,33 +116,62 @@ def construct_client(
             raise RuntimeError(f"Unexpected transport type: {transport_name}")
 
-        return client_class(transport=transport)
+        # return client_class(transport=transport)
```
Remove commented code

parthea commented on tests/system/conftest.py:
```diff
@@ -107,33 +116,62 @@ def construct_client(
     else:
         transport_cls = client_class.get_transport_class(transport_name)
         if transport_name in ["grpc", "grpc_asyncio"]:
+            # TODO: Need to test grpc transports without a channel_creator
```
Add link to bug for TODO comment . Applies throughout.

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -316,36 +322,110 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+        return universe_domain
+
+    @staticmethod
+    def _compare_universes(client_universe: str,
```
Add type hint for return value

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:
```diff
@@ -172,22 +184,42 @@ class {{ service.async_client_name }}:
+                2. If the GOOGLE_API_USE_CLIENT_CERTIFICATE environment variable
                 is "true", then the ``client_cert_source`` property can be used
-                to provide client certificate for mutual TLS transport. If
+                to provide client certificate for mTLS transport. If
```
```suggestion
                to provide a client certificate for mTLS transport. If
```

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -369,22 +449,32 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
+                2. If the GOOGLE_API_USE_CLIENT_CERTIFICATE environment variable
                 is "true", then the ``client_cert_source`` property can be used
-                to provide client certificate for mutual TLS transport. If
+                to provide client certificate for mTLS transport. If
```
```suggestion
                to provide a client certificate for mTLS transport. If
```

parthea commented on gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:
```diff
@@ -369,22 +449,32 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 set, no client certificate will be used.
+                
+                3. The ``universe_domain`` property can be used to override the
+                default "googleapis.com" universe. Note that ``api_endpoint``
```
```suggestion
                default "googleapis.com" universe. Note that the ``api_endpoint``
```

parthea reviewed (approved):
LGTM. Added minor observations

ohmayr committed:
only run tests with channel creation for grpc

ohmayr committed:
fix pytest parameters

ohmayr committed:
pass parameters as a single string

ohmayr committed:
pass credentials to transport

ohmayr committed:
pass credentials directly via transport

ohmayr committed:
remove credentials from client instance

ohmayr committed:
update test case to not directly pass in credentials

ohmayr committed:
remove async from parameter list

ohmayr committed:
address PR comments

ohmayr committed:
make parameters optional for _get_universe_domain

ohmayr committed:
update function parameter types

ohmayr committed:
use pytest to parameterize values

ohmayr committed:
fix style check

ohmayr committed:
Merge branch 'main' into implement-universe-domain
